internal/controller/maintenancereplica_controller.go

```diff
@@ -0,0 +1,217 @@
+// the ServerMaintenanceReplica object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile
+func (r *ServerMaintenanceReplicaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	replica := &metalv1alpha1.ServerMaintenanceReplica{}
+
+	if err := r.Get(ctx, req.NamespacedName, replica); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.reconcileExists(ctx, replica)
+}
+
+func (r *ServerMaintenanceReplicaReconciler) reconcileExists(ctx context.Context, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	if !replica.DeletionTimestamp.IsZero() {
+		return r.delete(ctx, log, replica)
+	}
+
+	if modified, err := clientutils.PatchEnsureFinalizer(ctx, r.Client, replica, ServerMaintenanceReplicaFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Ensured finalizer has been added")
+
+	return r.reconcile(ctx, log, replica)
+}
+
+func (r *ServerMaintenanceReplicaReconciler) delete(ctx context.Context, log logr.Logger, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log.V(1).Info("Deleting ServerMaintenanceReplica")
+
+	log.V(1).Info("Ensuring that the finalizer is removed")
+	if modified, err := clientutils.PatchEnsureNoFinalizer(ctx, r.Client, replica, ServerMaintenanceReplicaFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Deleted ServerMaintenanceReplica")
+	return ctrl.Result{}, nil
+}
+
+func (r *ServerMaintenanceReplicaReconciler) reconcile(ctx context.Context, log logr.Logger, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log.V(1).Info("Reconciling ServerMaintenanceReplica")
+	servers, err := r.getServersBySelector(ctx, replica)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(servers.Items) == 0 {
+		log.V(1).Info("No servers found")
+		return ctrl.Result{}, nil
+	}
+
+	maintenancelist := &metalv1alpha1.ServerMaintenanceList{}
+	if err = r.List(ctx, maintenancelist, client.MatchingLabels{
+		ServerMaintenanceReplicaFinalizer: replica.Name,
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If all servers have a maintenance object, only update the status
+	if len(servers.Items) == len(maintenancelist.Items) {
+		if modified, err := r.patchStatus(ctx, replica, calculateStatus(replica, maintenancelist.Items)); err != nil || modified {
+			return ctrl.Result{}, err
+		}
+	}
+
+out:
+	for _, server := range servers.Items {
+		log.V(1).Info("Reconciling server", "server", server.Name)
+		// Check if the server already got a maintenance
+		for _, maintenance := range maintenancelist.Items {
+			if maintenance.Spec.ServerRef.Name == server.Name {
```

Comment:
Quadratic worst case runtime. Consider creating a `map[string]ServerMaintenance`, where the key is the server name, for fast lookups.

A `FieldIndexer` from controllerruntime might also be an option.